coil-network/src/commonMain/kotlin/coil3/network/utils.common.kt

```diff
@@ -0,0 +1,73 @@
+package coil3.network
+
+import coil3.disk.DiskCache
+import io.ktor.http.HeadersBuilder
+import io.ktor.util.StringValues
+import io.ktor.utils.io.ByteReadChannel
+import io.ktor.utils.io.core.isEmpty
+import io.ktor.utils.io.core.readFully
+import okio.BufferedSink
+import okio.Closeable
+
+internal fun HeadersBuilder.append(line: String) = apply {
+    val index = line.indexOf(':')
+    require(index != -1) { "Unexpected header: $line" }
+    append(line.substring(0, index).trim(), line.substring(index + 1))
+}
+
+internal fun HeadersBuilder.appendAllIfNameAbsent(stringValues: StringValues) = apply {
+    stringValues.entries().forEach { (name, values) ->
+        if (!contains(name)) {
+            appendAll(name, values)
+        }
+    }
+}
+
+internal fun Closeable.closeQuietly() {
+    try {
+        close()
+    } catch (e: RuntimeException) {
+        throw e
+    } catch (_: Exception) {}
+}
+
+internal fun DiskCache.Editor.abortQuietly() {
+    try {
+        abort()
+    } catch (_: Exception) {}
+}
+
+/** Write a [ByteReadChannel] to [sink] using streaming. */
+internal suspend fun ByteReadChannel.writeTo(sink: BufferedSink) {
+    val buffer = ByteArray(OKIO_BUFFER_SIZE)
+
+    while (!isClosedForRead) {
+        val packet = readRemaining(buffer.size.toLong())
+        if (packet.isEmpty) break
+
+        // TODO: Figure out how to remove 'buffer' and read directly into 'sink'.
+        val bytesRead = packet.remaining.toInt()
```

Comment:
On JVM/Android, we can use ByteReadChannel.copyTo
https://github.com/ktorio/ktor/blob/main/ktor-io/jvm/src/io/ktor/utils/io/jvm/nio/Writing.kt
But I'm not sure if this API is subject to change

Reply:
Nice! Updated

Reply:
I wonder if it's possible to copy that extension's implementation and modify it to write to a `BufferedSink` directly. I took a quick look, but didn't look too closely

Reply:
> Nice! Updated

Aha, I suggest using FileChannel not BufferedSink
Since ByteReadChannel use DirectByteBuffer as inner buffer for now(current behaviour, not sure after ktor migrated to kotlinx-io)

Reply:
> I wonder if it's possible to copy that extension's implementation and modify it to write to a `BufferedSink` directly. I took a quick look, but didn't look too closely

Not possible, it rely on ByteReadChannel's platform implementation

Reply:
Another things to clarify, the ByteReadChannel. not always provide DirectByteBuffer, various by engine.

https://github.com/FooIbar/EhViewer/blob/main/app/src/oss/kotlin/com/hippo/ehviewer/ktor/CronetEngine.kt
Here I wrapped Android 14 HttpEngine(actually it's cronet)as a ktor engine, it always provide DirectByteBuffer as Cronet use DirectByteBuffer.
Anyway, engine provides HeapByteBuffer, we won't lose performance using FileChannel, engine provides DirectByteBuffer, we gain performance for zero-copying

Reply:
Gotcha - reverted the previous commit. Feel free to send a PR that uses `FileChannel`!

Reply:
> Gotcha - reverted the previous commit. Feel free to send a PR that uses `FileChannel`!

With nonJvm platforms, we can use UnsafeCursor(https://square.github.io/okio/3.x/okio/okio/okio/-buffer/-unsafe-cursor/index.html) though, I don't suggest to do this